discovery.go

```diff
@@ -0,0 +1,368 @@
+		advertiseCancel()
+		delete(d.advertising, topic)
+	}
+}
+
+// Bootstrap handles initial bootstrapping of peers for a given topic
+func (d *discover) Bootstrap(async bool, topic string, dOpts ...discovery.Option) {
+	if d.discovery == nil {
+		return
+	}
+
+	d.bootstrappedMux.RLock()
+	doneCh, bootstrapStarted := d.bootstrapped[topic]
+	d.bootstrappedMux.RUnlock()
+
+	if bootstrapStarted {
+		if !async {
+			_ = <-doneCh
+		}
+	} else {
+		d.bootstrappedMux.Lock()
+		doneCh, bootstrapStarted := d.bootstrapped[topic]
+		if bootstrapStarted {
+			d.bootstrappedMux.Unlock()
+			if !async {
+				_ = <-doneCh
+			}
+			return
+		}
+		ch := make(chan struct{}, 1)
+		d.bootstrapped[topic] = ch
+		d.bootstrappedMux.Unlock()
+
+		if async {
+			go func() {
+				d.handleDiscovery(d.p.ctx, topic, dOpts)
+				ch <- struct{}{}
+				close(ch)
+			}()
+		} else {
+			d.handleDiscovery(d.p.ctx, topic, dOpts)
+			ch <- struct{}{}
+			close(ch)
+		}
+	}
+}
+
+// Discover searches for additional peers interested in a given topic
+func (d *discover) Discover(topic string, opts ...discovery.Option) {
+	if d.discovery == nil {
+		return
+	}
+
+	d.discoverQ <- &discoverReq{topic, opts}
+}
+
+func (d *discover) handleDiscovery(ctx context.Context, topic string, opts []discovery.Option) {
+	discoverCtx, cancel := context.WithTimeout(ctx, time.Second*10)
+	defer cancel()
+
+	peerChan, err := d.discovery.FindPeers(discoverCtx, topic, opts...)
+	if err != nil {
+		log.Debugf("error finding peers for topic %s: %v", topic, err)
+		return
+	}
+
+	findPeers := make(map[peer.ID]struct{})
+
+loop:
+	for {
+		select {
+		case pi, more := <-peerChan:
+			if pi.ID == d.p.host.ID() || pi.ID == "" {
+				if more {
+					continue
+				} else {
+					break loop
+				}
+			}
+
+			findPeers[pi.ID] = struct{}{}
+
+			go func(pi peer.AddrInfo) {
+				ctx, cancel := context.WithTimeout(ctx, time.Second*10)
+				defer cancel()
+
+				err := d.p.host.Connect(ctx, pi)
+				if err != nil {
+					log.Debugf("Error connecting to pubsub peer %s: %s", pi.ID, err.Error())
+					return
+				}
+			}(pi)
+
+			if !more {
+				break loop
+			}
+
+		case <-discoverCtx.Done():
+			log.Infof("pubsub discovery for %s timeout", topic)
+			return
+		}
+	}
+
+	if len(findPeers) > 0 {
+		req := &reqNewPeerNotifier{topic: topic, resp: make(chan chan peer.ID, 1)}
+		d.addPeerNotif <- req
+		peerCh := <-req.resp
+		defer func() { d.rmPeerNotif <- peerCh }()
+		for {
+			select {
+			case pid := <-peerCh:
+				if _, ok := findPeers[pid]; ok {
+					delete(findPeers, pid)
+					log.Debugf("Connected to pubsub peer %s", pid)
+					if len(findPeers) == 0 {
+						return
+					}
+				}
+			case <-discoverCtx.Done():
+				log.Infof("pubsub discovery for %s timeout", topic)
+				return
+			}
+		}
+	}
+}
+
+func (d *discover) NotifyNewPeer(pid peer.ID) {
+	if d.discovery == nil {
+		return
+	}
+
+	select {
+	case d.peerNotifier <- pid:
+	default:
+		log.Error("dropped msg")
```

Comment:
better log here, what message did you drop?

Reply:
dropped the peer discovery notification message, will make the log message more detailed.